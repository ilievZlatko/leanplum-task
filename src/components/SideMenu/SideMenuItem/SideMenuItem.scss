@import '../../../styles/spacings';
@import '../../../styles/colors';
@import '../../../styles/mixins';

.menu-item {
	padding: $space-lg;
	border: 2px solid transparent;
	@include border-radius;
	cursor: pointer;
	user-select: none;

	&__active {
		border: 2px solid $color-dark-text;
	}
}
