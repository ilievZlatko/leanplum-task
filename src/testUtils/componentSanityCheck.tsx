import React from 'react'
import { shallow } from 'enzyme'

export const componentSanityCheck = (
	label: string,
	Element: React.ReactElement,
) =>
	describe(`Sanity checks for ${label}`, () => {
		it('is defined', () => {
			expect(Element).toBeDefined()
		})

		it('renders the element', () => {
			const wrapper = shallow(Element)
			const expected = false
			const actual = wrapper.isEmptyRender()
			expect(expected).toEqual(actual)
		})

		it('matches the snapshot', () => {
			const wrapper = shallow(Element)
			const expected = wrapper.html()
			expect(expected).toMatchSnapshot()
		})
	})
