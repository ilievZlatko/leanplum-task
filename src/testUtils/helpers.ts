import { ShallowWrapper, ReactWrapper } from 'enzyme'

export const findByTestAttr = (
	component: ShallowWrapper | ReactWrapper | any,
	attr: string,
): ShallowWrapper | ReactWrapper => {
	return component.find(`[data-test='${attr}']`)
}
