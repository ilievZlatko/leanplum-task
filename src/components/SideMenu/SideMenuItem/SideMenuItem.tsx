import React from 'react'
import { UserEntity } from '../../../models/UserEntity'
import { TitleSubtitle } from '../../../components'
import './SideMenuItem.scss'

interface Props {
	user: UserEntity
	active: boolean | null
	onClick: (user: UserEntity) => void
}

const SideMenuItem: React.FC<Props> = ({ user, active, onClick }) => {
	return (
		<aside
			className={['menu-item', active ? 'menu-item__active' : '']
				.join(' ')
				.trim()}
			onClick={() => onClick(user)}
		>
			<TitleSubtitle
				title={user.id}
				subtitle={
					<>
						<span>{user.devices} Devices - </span>
						<span>{user.sessions.length} Sessions - </span>
						<span>{user.location}</span>
					</>
				}
			/>
		</aside>
	)
}

export default SideMenuItem
