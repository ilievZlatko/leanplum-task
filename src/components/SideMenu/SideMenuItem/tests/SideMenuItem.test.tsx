import React from 'react'
import { shallow } from 'enzyme'

import SideMenuItem from '../SideMenuItem'
import { componentSanityCheck } from '../../../../testUtils/componentSanityCheck'
import { users } from '../../tests/mocks'
import { findByTestAttr } from '../../../../testUtils/helpers'

describe('SideMenuItem', () => {
	componentSanityCheck(
		'SideMenuItem',
		<SideMenuItem user={users[0]} active={false} onClick={jest.fn()} />,
	)

	it('Should respond to click events, and pass the correct info back', (done: jest.DoneCallback) => {
		const onClick = jest.fn()
		const expectedEvent = users[4]

		const wrapper = shallow(
			<SideMenuItem user={users[4]} active={true} onClick={onClick} />,
		)
		const item = findByTestAttr(wrapper, 'menu-item')

		item.simulate('click')

		expect(onClick).toHaveBeenCalledWith(expectedEvent)
		done()
	})

	it('Should be marked as active if active property is set to true', (done: jest.DoneCallback) => {
		const wrapperActive = shallow(
			<SideMenuItem user={users[3]} active={true} onClick={jest.fn()} />,
		)

		const wrapperInactive = shallow(
			<SideMenuItem user={users[3]} active={false} onClick={jest.fn()} />,
		)

		expect(wrapperActive.hasClass('menu-item')).toEqual(true)
		expect(wrapperActive.hasClass('menu-item__active')).toEqual(true)
		expect(wrapperInactive.hasClass('menu-item')).toEqual(true)
		expect(wrapperInactive.hasClass('menu-item__active')).toEqual(false)
		done()
	})
})
