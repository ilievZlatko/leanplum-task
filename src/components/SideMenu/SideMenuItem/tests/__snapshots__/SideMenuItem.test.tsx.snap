// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenuItem Sanity checks for SideMenuItem matches the snapshot 1`] = `"<aside class=\\"menu-item\\" data-test=\\"menu-item\\"><div><h2 data-test=\\"title\\" class=\\"title\\">usr_0</h2><span data-test=\\"subtitle\\" class=\\"subtitle\\"><span>1 Devices - </span><span>2 Sessions - </span><span>Sofia</span></span></div></aside>"`;
